### PR TITLE
CCTP volume tracking

### PR DESCRIPTION
**Pull request by d4v1d03**

### CODE REVIEW

File: Bridge/cctp/src/handlers/MessageReceived.ts [hunk 1]
```diff
@@ -123,4 +146,10 @@ export const MessageReceivedHandler = async (
       id: feeinUSDId,
       feeInUSDC: feeamount,
     });
+
+  getTodayEntry(block.chain_id, todayEntryDB, amount, feeamount);
+  getWeeklyEntry(block.chain_id, weekEntryDB, amount, feeamount);
+  getMonthlyEntry(block.chain_id, monthEntryDB, amount, feeamount);
+  getYearlyEntry(block.chain_id, yearEntryDB, amount, feeamount);
+  getAllTimeEntry(block.chain_id, allTimeEntryDB, amount, feeamount);
```
Comment (pradyuman-verma) [inline Bridge/cctp/src/handlers/MessageReceived.ts:154]: these are async operations, need to add await.

File: Bridge/cctp/src/handlers/DepositForBurn.ts [hunk 2]
```diff
@@ -33,12 +53,21 @@ export const DepositForBurnHandler = async (
...
+  const weekEntryDB: Instance = bind(cctpWeekDataDB);
+  const monthEntryDB: Instance = bind(cctpMonthDataDB);
+  const yearEntryDB: Instance = bind(cctpYearDataDB);
+  const allTimeEntryDB: Instance = bind(cctpAllTimeDB);
+
+  getTodayEntry(block.chain_id, todayEntryDB, amount, 0);
+  getWeeklyEntry(block.chain_id, weekEntryDB, amount, 0);
+  getMonthlyEntry(block.chain_id, monthEntryDB, amount, 0);
+  getYearlyEntry(block.chain_id, yearEntryDB, amount, 0);
+  getAllTimeEntry(block.chain_id, allTimeEntryDB, amount, 0);
```
Comment (pradyuman-verma) [inline Bridge/cctp/src/handlers/DepositForBurn.ts:70]: await

File: Bridge/cctp/src/utils/tracking.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    withdrawal: "0",
+    feeTotal: "0",
+  });
+  entry.txCount = new BigNumber(entry.txCount).plus(1).toString();
+  entry.dailyVolume = new BigNumber(entry.dailyVolume).plus(amount).toString();
+  if (totalFee) {
+    entry.deposited = new BigNumber(entry.deposited).plus(amount).toString();
+  } else {
+    entry.withdrawal = new BigNumber(entry.withdrawal).plus(amount).toString();
+  }
```
Comment (pradyuman-verma) [inline Bridge/cctp/src/utils/tracking.ts:33]: entry.totalFee tracking ??

File: Bridge/cctp/src/utils/tracking.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
+import { Instance } from "@blockflow-labs/utils";
+import BigNumber from "bignumber.js";
+
+const today = new Date();
+const date = today.toISOString().split("T")[0];
+const week = Math.floor(Date.now() / 604800000);
+const month = new Date().getMonth();
+const year = new Date().getFullYear();
+
```
Comment (pradyuman-verma) [inline Bridge/cctp/src/utils/tracking.ts:9]: We can't use this to calculate, we need to use block.timestamp for all calculations. 

For example, if an event happened last month CCTP, we are processing and counting it in today timestamp.

File: Bridge/cctp/src/handlers/MessageReceived.ts [hunk 5]
```diff
@@ -123,4 +146,21 @@ export const MessageReceivedHandler = async (
       id: feeinUSDId,
       feeInUSDC: feeamount,
     });
+
+  const blockTimestamp = Number(block.block_timestamp);
+  const dateFromTimestamp = new Date(blockTimestamp* 1000);
+  const date = dateFromTimestamp.toISOString().split("T")[0];
+  const week = Math.floor(dateFromTimestamp.getTime() / 604800000);
+  const month = dateFromTimestamp.getMonth() + 1; 
+  const year = dateFromTimestamp.getFullYear();
```
Comment (pradyuman-verma) [inline Bridge/cctp/src/handlers/MessageReceived.ts:155]: why calculating it here? Just pass timestamp as a param to all functions.

Comment (d4v1d03) [reply Bridge/cctp/src/handlers/MessageReceived.ts]: fixed it

**Comment by d4v1d03** [comment]
image
tests working fine

**Comment by d4v1d03** [comment]
DB's saving data 
image

File: Bridge/cctp/src/utils/tracking.ts [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    withdrawal: "0",
+    feeTotal: "0",
+  });
+  entry.txCount = new BigNumber(entry.txCount).plus(1).toString();
+  entry.yearlyVolume = new BigNumber(entry.yearlyVolume)
+    .plus(amount)
+    .toString();
+  if (totalFee) {
+    entry.deposited = new BigNumber(entry.deposited).plus(amount).toString();
+    entry.feeTotal = new BigNumber(entry.feeTotal).plus(totalFee).toString();
```
Comment (pradyuman-verma) [inline Bridge/cctp/src/utils/tracking.ts:136]: It's `totalFee` in schema.

Comment (pradyuman-verma) [reply Bridge/cctp/src/utils/tracking.ts]: Fixed

**Review by pradyuman-verma — APPROVED** [review]

**Pull request merged**